java/src/main/java/ai/rapids/cudf/Table.java

```diff
@@ -1727,52 +1745,60 @@ public Table aggregate(Aggregate... aggregates) {
-        for (AggregateOp operation: entry.getValue().operations()) {
-          aggColumnIndexes[opIndex] = columnIndex;
-          aggOperationIds[opIndex] = operation.nativeId;
-          opIndex++;
-        }
-      }
-      assert opIndex == totalOps: opIndex + " == " + totalOps;
-
-      Table aggregate;
-      aggregate = new Table(groupByAggregate(
-          operation.table.nativeHandle,
-          operation.indices,
-          aggColumnIndexes,
-          aggOperationIds,
-          groupByOptions.getIgnoreNullKeys()));
+      long[] aggOperationInstances = new long[totalOps];
       try {
-        // prepare the final table
-        ColumnVector[] finalCols = new ColumnVector[keysLength + aggregates.length];
-
-        // get the key columns
-        for (int aggIndex = 0; aggIndex < keysLength; aggIndex++) {
-          finalCols[aggIndex] = aggregate.getColumn(aggIndex);
+        int opIndex = 0;
+        for (Map.Entry<Integer, ColumnOps> entry: groupedOps.entrySet()) {
+          int columnIndex = entry.getKey();
+          for (Aggregation operation: entry.getValue().operations()) {
+            aggColumnIndexes[opIndex] = columnIndex;
+            aggOperationInstances[opIndex] = operation.createNativeInstance();
+            opIndex++;
+          }
         }
+        assert opIndex == totalOps : opIndex + " == " + totalOps;
+
+        Table aggregate;
+        aggregate = new Table(groupByAggregate(
+                operation.table.nativeHandle,
+                operation.indices,
+                aggColumnIndexes,
+                aggOperationInstances,
+                groupByOptions.getIgnoreNullKeys()));
+        try {
+          // prepare the final table
+          ColumnVector[] finalCols = new ColumnVector[keysLength + aggregates.length];
+
+          // get the key columns
+          for (int aggIndex = 0; aggIndex < keysLength; aggIndex++) {
+            finalCols[aggIndex] = aggregate.getColumn(aggIndex);
+          }
 
-        int inputColumn = keysLength;
-        // Now get the aggregation columns
-        for (ColumnOps ops: groupedOps.values()) {
-          for (List<Integer> indices: ops.outputIndices()) {
-            for (int outIndex: indices) {
-              finalCols[outIndex] = aggregate.getColumn(inputColumn);
+          int inputColumn = keysLength;
+          // Now get the aggregation columns
+          for (ColumnOps ops: groupedOps.values()) {
+            for (List<Integer> indices: ops.outputIndices()) {
+              for (int outIndex: indices) {
+                finalCols[outIndex] = aggregate.getColumn(inputColumn);
+              }
+              inputColumn++;
             }
-            inputColumn++;
           }
+          return new Table(finalCols);
+        } finally {
+          aggregate.close();
         }
-        return new Table(finalCols);
       } finally {
-        aggregate.close();
+        for (long instance : aggOperationInstances) {
```

Comment:
Suggestion: This cleanup pattern occurs a number of times, might be worth a private utility function to help reduce the error path boilerplate a bit.